code/modules/hydroponics/compost_bin.dm

```diff
@@ -0,0 +1,171 @@
+/obj/machinery/compost_bin/attackby(obj/item/O, mob/user, params)
+	if(user.a_intent == INTENT_HARM)
+		return ..()
+
+	if(istype(O, /obj/item/storage/bag/plants))
+		if(biomass >= biomass_capacity)
+			to_chat(user, "<span class='warning'>[src] can't hold any more biomass!</span>")
+			return
+
+		var/obj/item/storage/bag/plants/PB = O
+		for(var/obj/item/reagent_containers/food/snacks/grown/G in PB.contents)
+
+			if(biomass >= biomass_capacity)
+				break
+			PB.remove_from_storage(G, src)
+			make_biomass(G)
+
+		if(biomass < biomass_capacity)
+			to_chat(user, "<span class='info'>You empty [PB] into [src].</span>")
+		else
+			to_chat(user, "<span class='info'>You fill [src] to its capacity.</span>")
+
+		SStgui.update_uis(src)
+		update_icon_state()
+		return TRUE
+
+	if(istype(O, /obj/item/reagent_containers/food/snacks/grown))
+		if(biomass >= biomass_capacity)
+			to_chat(user, "<span class='warning'>[src] can't hold any more plants!</span>")
+			return
+		if(!user.unEquip(O))
+			return
+
+		O.forceMove(src)
+		make_biomass(O)
+		to_chat(user, "<span class='info'>You put [O] in [src].</span>")
+		SStgui.update_uis(src)
+		update_icon_state()
+		return TRUE
+
+	to_chat(user, "<span class='warning'>You cannot put this in [name]!</span>")
+
+/obj/machinery/compost_bin/attack_hand(mob/user)
+	ui_interact(user)
+
+
+/obj/machinery/compost_bin/ui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = FALSE, datum/tgui/master_ui = null, datum/ui_state/state = GLOB.default_state)
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "CompostBin", "CompostBin", 390, 200, master_ui, state)
+		ui.set_autoupdate(FALSE)
+		ui.open()
+
+
+/obj/machinery/compost_bin/ui_data(mob/user)
+	var/list/data = list()
+	data["biomass"] = biomass
+	data["biomass_capacity"] = biomass_capacity
+	data["compost"] = compost
+	data["compost_capacity"] = compost_capacity
+	return data
+
+//Compost compostable material if there is any
+/obj/machinery/compost_bin/process()
+	if(compost >= compost_capacity || biomass <= 0)
+		return
+	process_counter++
+	if(process_counter < 5)
+		return
+	process_counter = 0
+	//converts 20% of the biomass to compost each cycle, unless there isn't enough compost space or there is 10 or less biomass
+	var/conversion_amount = clamp(DECAY * biomass, min(MIN_CONVERSION, biomass), compost_capacity - compost)
+	biomass -= conversion_amount
+	compost += conversion_amount
+	update_icon_state()
+	SStgui.update_uis(src)
+
+// Makes soil from compost
+/obj/machinery/compost_bin/proc/create_soil(amount)
+	// Verify theres enough compost
+	if(compost < (SOIL_COST * amount))
+		return
+	new /obj/item/stack/sheet/soil(loc, amount)
+	compost -= SOIL_COST * amount
+	update_icon_state()
+	SStgui.update_uis(src)
+
+// calls functions according to ui interaction(just making compost for now)
+/obj/machinery/compost_bin/ui_act(action, list/params)
```

Comment:
can you move the other two `ui_x` procs down so that all of the UI procs are grouped together?